src/jdk.naming.dns/share/classes/com/sun/jndi/dns/DnsClient.java

```diff
@@ -402,43 +428,72 @@ private byte[] doUdpQuery(Packet pkt, InetAddress server,
                                      int port, int retry, int xid)
             throws IOException, NamingException {
 
-        int minTimeout = 50; // msec after which there are no retries.
-
-        synchronized (udpSocketLock) {
-            try (DatagramSocket udpSocket = getDatagramSocket()) {
-                DatagramPacket opkt = new DatagramPacket(
-                        pkt.getData(), pkt.length(), server, port);
-                DatagramPacket ipkt = new DatagramPacket(new byte[8000], 8000);
+        udpChannelLock.lock();
+        try {
+            try (DatagramChannel udpChannel = getDatagramChannel()) {
+                ByteBuffer opkt = ByteBuffer.wrap(pkt.getData(), 0, pkt.length());
+                byte[] data = new byte[8000];
+                ByteBuffer ipkt = ByteBuffer.wrap(data);
                 // Packets may only be sent to or received from this server address
-                udpSocket.connect(server, port);
+                InetSocketAddress target = new InetSocketAddress(server, port);
+                udpChannel.connect(target);
                 int pktTimeout = (timeout * (1 << retry));
-                udpSocket.send(opkt);
+                udpChannel.send(opkt, target);
```

Comment:
FYI: When a DatagramChannel is connected then you can use `write` to send the datagram, you don't need to specify the target address again when sending.